server/src/com/cloud/template/TemplateManagerImpl.java

```diff
@@ -509,7 +509,7 @@ private String extract(Account caller, Long templateId, String url, Long zoneId,
             for (DataStore store : ssStores) {
                 tmpltStoreRef = _tmplStoreDao.findByStoreTemplate(store.getId(), templateId);
                 if (tmpltStoreRef != null) {
-                    if (tmpltStoreRef.getDownloadState() == com.cloud.storage.VMTemplateStorageResourceAssoc.Status.DOWNLOADED) {
+                    if (tmpltStoreRef.getDownloadState() == com.cloud.storage.VMTemplateStorageResourceAssoc.Status.DOWNLOADED && tmpltStoreRef.getDownloadPercent() == 100) {
```

Comment:
@yvsubhash thanks for the patch.

Can we address this (add this check) while setting the download state to DOWNLOADED? That way unless download progress is 100%, the template would not move to DOWNLOADED state.
Also good add a unit here, checking if template download progress is 100 if template state is DOWNLOADED.